slick/src/main/scala/slick/jdbc/LoggingStatement.scala

```diff
@@ -1,20 +1,89 @@
 package slick.jdbc
 
+import slick.util.TableDump
+
+import scala.collection.mutable.ArrayBuffer
 import scala.language.reflectiveCalls
 
-import java.sql.{PreparedStatement, Connection, SQLWarning, ResultSet, Statement}
+import java.io.{InputStream, Reader}
+import java.util.Calendar
+import java.{sql => js}
+import java.sql.{PreparedStatement, Connection, SQLWarning, ResultSet, Statement, Timestamp}
 
 /** A wrapper for `java.sql.Statement` that logs statements and benchmark results
   * to the appropriate [[JdbcBackend]] loggers. */
 class LoggingStatement(st: Statement) extends Statement {
   private[this] val doStatement = JdbcBackend.statementLogger.isDebugEnabled
   private[this] val doBenchmark = JdbcBackend.benchmarkLogger.isDebugEnabled
+  private[this] val doParameter = JdbcBackend.parameterLogger.isDebugEnabled
+
+  private[this] var params: ArrayBuffer[(Any, Any)] = null
+  private[this] var paramss: ArrayBuffer[ArrayBuffer[(Any, Any)]] = null
+
+  /** log a parameter */
+  protected[this] def p(idx: Int, tpe: Any, value: Any): Unit = if(doParameter) {
+    if(params eq null) params = new ArrayBuffer
+    if(params.size == idx) params += ((tpe, value))
+    else {
+      while(params.size < idx) params += null
+      params(idx-1) = (tpe, value)
+    }
+  }
 
-  @inline protected[this] def logged[T](sql: String, what: String = "statement")(f: =>T) = {
+  protected[this] def pushParams: Unit = if(doParameter) {
```

Comment:
This distinction is not made consistently in other parts of the codebase. IMHO it is not very useful, either. It only applies to nullary methods, you often cannot rely on it with Scala code and never with Java code.